netwaystev2/src/filter/sortedbuffer.rs

```diff
@@ -0,0 +1,251 @@
+///! A sorted buffer which maintains ordering and uniqueness in the buffer.
+
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+
+pub struct SequencedMinHeap<T> {
+    heap: BinaryHeap<Reverse<SequencedItem<T>>>,
+}
+
+impl<T> SequencedMinHeap<T> {
+    pub fn new() -> Self {
+        Self {
+            heap: BinaryHeap::new(),
+        }
+    }
+
+    fn contains_sequence_number(&self, sequence: u64) -> bool {
+        for tuple in &self.heap {
+            if tuple.0.0 == sequence {
+                return true
+            }
+        }
+        false
+    }
+
+    pub fn count_contiguous(&self, mut sequence: u64) -> usize {
+        let mut count = 0;
+        for tuple in &self.heap {
+            if tuple.0.0 == sequence {
+                count += 1;
+                sequence += 1;
+            }
+        }
+        return count;
+    }
+
+    /// Add this T to the sequenced min-heap. Returns false if not added because
+    /// this sequence is already in the min-heap; otherwise, returns true.
+    pub fn add(&mut self, sequence: u64, t: T) -> bool {
+        if self.contains_sequence_number(sequence) {
+            return false;
+        }
+        self.heap.push(Reverse(SequencedItem(sequence, t)));
+        true
+    }
+
+    /// Gets the minimum sequence number in the min-heap. This is the sequence number of
+    /// what we would .take()
+    pub fn peek_sequence_number(&self) -> Option<u64> {
+        self.heap.peek().map(|reversed_tup| reversed_tup.0.0)
+    }
+
+    /// Takes the T with the lowest sequence number
+    pub fn take(&mut self) -> Option<T> {
+        self.heap.pop().map(|reversed_tup| reversed_tup.0.1)
+    }
+
+    #[cfg(test)]
+    pub fn print(&self) {
+        for tuple in &self.heap {
+            println!("Key: {}", tuple.0.0);
+        }
+    }
+}
+
+struct SequencedItem<T>(u64, T);
+
+impl<T> Ord for SequencedItem<T> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl<T> PartialEq for SequencedItem<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<T> PartialOrd for SequencedItem<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// Marker trait impl for Ord without needing to pass `Eq` down to the item
+impl<T> Eq for SequencedItem<T> {
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
What do you think about moving all these tests to a separate file, and in general, keeping our tests separate from our business logic?

Reply:
Love the idea. We tend to write quite a bit of tests anyways which makes the files unmanageable during navigation.

Reply:
Needed to make one function public in order to access it. I tried `pub(crate)` first but that didn't work since tests are seen as external to the crate.

Reply:
Well... they don't have to be _that_ separate... you have them as integration tests: https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests

That said, it's not a big deal that `contains_sequence_number` is public now.

Reply:
I was thinking about this this morning and decided to go with `src/filter/tests` instead. Will be in latest set of commits once I push.